rust/cadeau/src/xmf/vpx/mod.rs

```diff
@@ -0,0 +1,225 @@
+    XmfVpxFrame_GetHeight, XmfVpxFrame_GetPartitionId, XmfVpxFrame_GetPts, XmfVpxFrame_GetSize,
+    XmfVpxFrame_GetSpatialLayerEncoded, XmfVpxFrame_GetWidth, XmfVpxImage, XmfVpxImage_Destroy, XmfVpxPacket,
+    XmfVpxPacket_Destroy, XmfVpxPacket_GetFrame, XmfVpxPacket_GetKind, XmfVpxPacket_IsEmpty,
+};
+
+pub mod decoder;
+pub mod encoder;
+
+#[derive(Debug, Clone, Copy)]
+pub enum VpxCodec {
+    VP8,
+    VP9,
+}
+
+pub struct VpxImage {
+    pub(crate) ptr: *const XmfVpxImage,
+}
+
+impl Drop for VpxImage {
+    fn drop(&mut self) {
+        // Safety: it is safe to call, the owenership of the pointer is managed by the XmfVpxImage itself.
+        unsafe {
+            XmfVpxImage_Destroy(self.ptr.cast_mut());
+        }
+    }
+}
+
+impl From<VpxCodec> for XmfVpxCodecType {
+    fn from(codec: VpxCodec) -> Self {
+        match codec {
+            VpxCodec::VP8 => XmfVpxCodecType::VP8,
+            VpxCodec::VP9 => XmfVpxCodecType::VP9,
+        }
+    }
+}
+
+pub fn is_key_frame(buffer: &[u8]) -> bool {
+    if buffer.is_empty() {
+        return false;
+    }
+    buffer[0] & 0x1 == 0
+}
+
+pub struct VpxPacket {
+    pub(crate) ptr: *const XmfVpxPacket,
+}
+
+impl VpxPacket {
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn kind(&self) -> XmfVpxPacketKind {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_GetKind(self.ptr.cast_mut()) }
+    }
+
+    /// # Safety
+    ///
+    /// Obtain the deep copy of the frame from the packet, must be used before any call to VpxEncoder method.
+    /// The frame produced by this method is always valid.
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn frame(&self) -> Option<VpxFrame> {
+        // Safety: it's up to the caller to ensure the packet is valid, the frame will remain valid until the frame itself is destroyed.
+        let frame_ptr = unsafe { XmfVpxPacket_GetFrame(self.ptr) };
+        if frame_ptr.is_null() {
+            None
+        } else {
+            VpxFrame::new(frame_ptr).ok()
+        }
+    }
+
+    /// # Safety
+    ///
+    /// This function is `unsafe` because it assumes that `self.ptr` is a valid pointer.
+    /// It should not be used after any call to a `VpxEncoder` method that modifies or releases the packet.
+    pub unsafe fn is_empty(&self) -> bool {
+        // Safety: It's up to the caller to ensure the packet is valid.
+        unsafe { XmfVpxPacket_IsEmpty(self.ptr) }
+    }
+}
+
+impl Drop for VpxPacket {
+    fn drop(&mut self) {
+        // Safety: Ensure that the pointer is valid and has not been freed before calling the drop function.
```

Comment:
issue: The comment should exactly explain _how_ we ensured that. This is stating a safety precondition. Safety preconditions should be in the documentation of unsafe functions. They can’t be used to prove that an unsafe call is safe.

Reply:
updated